### ISSUE TITLE
DataviewJS output (error)

### ISSUE THREAD

**Issue by dxcore35** [mdelobelle/metadatamenu#274]
Hello I have used this dataviewJS to create list of location from daily note:

# DataviewJS query code
```
const pages = dv.pages()
	     .where(b => b.location != '' && b.class == "daily-note")        
        .sort(b => b.file.mtime, "asc")
		.location

let uniqueValues = Array.from(new Set(pages));

console.log(uniqueValues)
dv.list(uniqueValues)
```
# DataviewJS query in note body
When adding this dataviewJS into note, it generates list of unique locations.

# DataviewJS query in Metadata menu

After copying it here:

![Screenshot 2023-02-01 at 17 11 24](https://user-images.githubusercontent.com/7279156/216081735-2d4471bf-6287-419b-9a4b-45ddb57d1189.jpg)

There is no output. 

---

What I don't understand what object you are reading when reading this query...

**Comment by auratheenby** [comment]
it's not trying to actually render dataview, it's just looking for which pages you want it to apply to. So no need for `sort()`, or telling dataview to render via `dv.list()`. 

You might also be interested to know that `DataArray` has a method called `distinct()`, which achieves the same thing as turning it into a set and back again.


This should work just fine:
```javascript
 dv.pages()
     .where(b => b.location != '' && b.class == "daily-note")
     .distinct()
 ```

**Comment by dxcore35** [comment]
The result of proposed query is returning also duplicates. I need to add `uniqueValues`. I don't understand your explanation. So what object I can use as input for Metadata menu dataview query?

**Comment by auratheenby** [comment]
@dxcore35 
Actually, looking at the query a bit more, I don't think you should be getting duplicates in the first place - a blank  `dv.pages()` should only be returning one copy of each page.

Also, just curious why you don't set the fileClass field alias to `class` or rename all the instances of `class` to `fileClass`? I suppose if you have an issue with the location being blank, but seems more like a metadata issue which would be better solved by using the table view. :shrug:

In terms of how the query works, I looked at the source:
https://github.com/mdelobelle/metadatamenu/blob/864d406e9227853796422d8330a1e1e5662f6c26/src/fileClass/FileClassQuery.ts#L15
It literally just appends your input onto a `return` statement. So just think of it being returned or assigned to a variable, I guess.

> I don't understand why there is no inbuilt function "Unique values from a dataview query". This will be so handy for so many use cases.

I mean, like I said, `distinct()` exists (even though it apparently doesn't work for you for some reason), but I _would_ love a keyword for in query language mode like WHERE or SORT.

**Comment by auratheenby** [comment]
Sorry, didn't read that whole thing quite right - I thought that was for the 'daily-note' `fileClass`. I guess I need to improve my reading comprehension :grimacing: 

I did a bit of testing in a fresh vault with your setup, and this works for me:
```javascript
dv.pages()
.where(b => b.location != '' && b.class == 'daily-note')
.location
.distinct()
```
About how you're supposed use these queries, I took a look at the source:
https://github.com/mdelobelle/metadatamenu/blob/864d406e9227853796422d8330a1e1e5662f6c26/src/fileClass/FileClassQuery.ts#L15
It literally just appends your input onto a `return` statement.

**Comment by dxcore35** [comment]
> Sorry, didn't read that whole thing quite right - I thought that was for the 'daily-note' `fileClass`. I guess I need to improve my reading comprehension 😬
> 
> I did a bit of testing in a fresh vault with your setup, and this works for me:
> 
> ```js
> dv.pages()
> .where(b => b.location != '' && b.class == 'daily-note')
> .location
> .distinct()
> ```
> 
> About how you're supposed use these queries, I took a look at the source:
> 
> https://github.com/mdelobelle/metadatamenu/blob/864d406e9227853796422d8330a1e1e5662f6c26/src/fileClass/FileClassQuery.ts#L15
> 
> 
> It literally just appends your input onto a `return` statement.

Thank you very much! This query works!

**Comment by auratheenby** [comment]
Glad I could help :)

**Comment by dxcore35** [comment]
**Related but expand question**

How to built the metadata that will take unique values (previous query) but also allow new value to be inserted.
With this query even in the editor mode when I type something it will gets deleted. As the Metadata menu class is forcing the values only match the dataview and remove all what is not matching.

**Comment by auratheenby** [comment]
@dxcore35 that tends to happen when you have auto updating on for the field. If you turn it off (even just temporarily) it should work. 

I would love the ability to just have the query as suggestions and not force the values. Maybe even adding new ones from the ui? 👀

**Comment by dxcore35** [comment]
I don't see auto updating option. Where is it?

**Comment by auratheenby** [comment]
In the field option menu, like the one you access when you’re setting the value query

**Comment by dxcore35** [comment]
@auratheenby Hi. I cannot find that field, could you please send me the screenshot where you see the option?

**Comment by auratheenby** [comment]
sorry, for some reason i thought there was an auto updating option with single values. not actually sure what it would do. :confused: anyways, i have encountered that before, but i don't remember how i fixed it, or if i even did anything to fix it other than give it some time and try again. sorry i'm not more helpful.
